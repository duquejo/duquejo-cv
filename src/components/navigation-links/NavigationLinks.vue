<template>
  <ul class="navigation w-full" :class="layoutClasses">
    <li v-for="route in ROUTES" :key="route" class="border-r border-gray-300 w-full text-center last:border-none lg:border-none lg:w-auto lg:text-right" :class="sizesClasses">
      <router-link :to="{ name: route }" @click="emit('link-click')">
        {{ generateMenuLink(route) }}
        <span v-if="layout === 'menu'">
          <v-icon class="ml-2" name="fa-chevron-right" scale="0.9" />
        </span>
      </router-link>
    </li>
  </ul>
</template>

<script setup lang="ts">
import { computed } from 'vue';
import { useI18n } from 'vue-i18n';

import { ROUTES } from '@/router';
import type { Sizes, NavigationLayout } from '@/interfaces';

interface Props {
  layout?: NavigationLayout;
  size?: Sizes
}

interface Emits {
  (e: 'link-click'): void;
}

const { t } = useI18n();

const emit = defineEmits<Emits>();
const props = withDefaults(defineProps<Props>(), {
  layout: 'menu',
  size: 'large',
});

const layoutClasses = computed(() => ({
  'invisible lg:visible text-right my-5 space-y-2  lg:p-0 px-8 py-10': props.layout === 'menu',
  'fixed flex bottom-0 bg-gray-100 justify-around px-4 pb-4 py-2.5 border-t border-gray-200 lg:hidden': props.layout === 'fixed',
}));

const sizesClasses = computed(() => ({
  'text-lg': props.size === 'large',
  'text-base': props.size === 'medium',
  'text-sm': props.size === 'small',
}));

const generateMenuLink = (routeName: ROUTES): string => t(`general.menu.${routeName}`);
</script>

<style scoped lang="postcss">

 /** Navigation links **/

 ul.navigation li a.router-link-active span {
  @apply opacity-100 decoration-yellow-400;
 }

 ul.navigation li a.router-link-active span svg {
  @apply fill-yellow-400;
 }

 ul.navigation li a.router-link-active {
  @apply underline underline-offset-8 decoration-yellow-400 decoration-2;
 }
 
 ul.navigation li:hover a {
  @apply text-inherit decoration-yellow-400;
 }

 ul.navigation li:hover a span {
  @apply opacity-70 ;
 }

 ul.navigation li a span {
  @apply opacity-0 transition-opacity;
 }
</style>