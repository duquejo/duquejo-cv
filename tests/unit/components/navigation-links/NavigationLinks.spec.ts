import { RouterLinkStub, shallowMount } from '@vue/test-utils';
import { beforeEach, describe, expect, it, vi } from 'vitest';

import NavigationLinks from '../../../../src/components/navigation-links/NavigationLinks.vue';
import createVueI18n from '../../../mocks/mock-i18n';


describe('+ NavigationLinks.vue tests', () => {

  const i18n = createVueI18n('en');

  const createWrapper = (
    layout = 'menu', 
    size = 'medium'
  ) => shallowMount(NavigationLinks, {
    props: {
      layout,
      size
    },
    global: {
      plugins: [i18n],
      stubs: {
        RouterLink: RouterLinkStub,
      }
    },
  });

  beforeEach(async () => {
    vi.clearAllMocks();
  });

  it('Should match the snapshot', () => {
    const wrapper = createWrapper();

    expect(wrapper.html()).toMatchSnapshot();
    
    const link = wrapper.findComponent(RouterLinkStub);
    const linkSpy = vi.spyOn(link, 'trigger');
    expect(linkSpy).not.toHaveBeenCalledOnce();

    expect(wrapper.html()).toContain('general.menu.');
  });

  it('should emit "click" event when RouterLink is clicked', async () => {

    const wrapper = createWrapper();

    const links = wrapper.findAllComponents(RouterLinkStub);
    
    links.forEach( async link => {
      const linkSpy = vi.spyOn(link, 'trigger');
      await link.trigger('click');
      expect(linkSpy).toHaveBeenCalled();
    });
    
    expect(wrapper.emitted()).toHaveProperty('click');
  });

  it('should load dynamic classes conditionally - menu', () => {
    const wrapper = createWrapper('menu', 'small');

    expect(wrapper.find('.navigation').classes()).toEqual([
      'navigation', 
      'w-full',
      'invisible', 
      'lg:visible', 
      'text-right', 
      'my-5',
      'space-y-2',
      'lg:p-0', 
      'px-8', 
      'py-10'
    ]);
    expect(wrapper.find('li').classes()).toContain('text-sm');
  });

  it('should load dynamic classes conditionally - fixed', () => {
    const wrapper = createWrapper('fixed', 'large');

    expect(wrapper.find('.navigation').classes()).toEqual([
      'navigation', 
      'w-full',
      'fixed',
      'flex',
      'bottom-0',
      'bg-gray-100',
      'justify-around',
      'px-4',
      'pb-4',
      'py-2.5',
      'border-t',
      'border-gray-200',
      'lg:hidden'
    ]);
    expect(wrapper.find('li').classes()).toContain('text-lg');
  });
});