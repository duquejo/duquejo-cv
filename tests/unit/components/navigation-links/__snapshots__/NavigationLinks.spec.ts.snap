// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`+ NavigationLinks.vue tests > Should match the snapshot 1`] = `
"<ul class="navigation w-full invisible lg:visible text-right my-5 space-y-2  lg:p-0 px-8 py-10" role="menu">
  <li class="border-r border-gray-300 w-full text-center last:border-none lg:border-none lg:w-auto lg:text-right text-base"><a>general.menu.resume <span><v-icon class="ml-2" name="bi-chevron-right" scale="0.9"></v-icon></span></a></li>
  <li class="border-r border-gray-300 w-full text-center last:border-none lg:border-none lg:w-auto lg:text-right text-base"><a>general.menu.services <span><v-icon class="ml-2" name="bi-chevron-right" scale="0.9"></v-icon></span></a></li>
</ul>"
`;
